### PR TITLE
[LTC] Lower convolution_overrideable to _convolution

### PR DESCRIPTION
**Pull request by desertfire**
Currently convolution_overrideable is implemented as a special fallback
due to the implementation limitations of convolution in Eager, which
brings two problems, 1) graph fragementation 2) fallback convolution
kernel may not be using the most efficient version for that device.
This PR fixes the forward convolution by lowering it to _convolution,
and it is verified to work for CPU and CUDA.

### CODE REVIEW

**Comment by wconstab** [comment]
this could be worth checking in as short term fix, but it only works for inference/forward right?  Have you thought about alternatives that also fix backward conv?

**Comment by wconstab** [comment]
@desertfire If you can confirm that this is working for backward (and explain how?) then we should definitely land it.  It seems to me like we should expect to catch a 'convolution_backward_overrideable' for the backward kernels if we go with this approach.  If that's what's happening, I would expect torchscript runtime to throw up its hands and complain that no real kernel is registered for device cuda.

**Comment by desertfire** [comment]
What TS IR is generated for convolution forward here is one level below LT, so convolution backward will still be captured as `convolution_backward_overrideable` by LT and the current backward fallback implementation in LT would still work. But I do want to push one step further by changing the backward IR generation as well. It will be a bit uglier because there is no unified `_convolution_backward`, but as a workaround, it will enable us to get the full graph for vision models.

**Comment by wconstab** [comment]
>  so convolution backward will still be captured as convolution_backward_overrideable by LT and the current backward fallback implementation in LT would still work.

I see.  This lines up with my expectation then.  We are not _really_ handling backwards, but unit tests pass because of fallback.   So this seems like it might be worth landing just to get folks unblocked with conv forward and inference models benchmarking.  But I am wary of trying to extend this approach to handle backwards, because I'm afraid it'll be pretty ugly and theres a cleaner fix (the lazy 'mode' idea) that we want to do anyway for other reasons.

> But I do want to push one step further by changing the backward IR generation as well. It will be a bit uglier because there is no unified _convolution_backward, but as a workaround, it will enable us to get the full graph for vision models.
Do you have an idea how to do this?

**Comment by desertfire** [comment]
@wconstab I have updated the implementation to lower backward using cudnn_convolution_backward. For CPU I have just left the current fallback approach untouched.

File: lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/aten_ltc_ts_type.cpp
Comment (wconstab) [inline lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/aten_ltc_ts_type.cpp:273]: is LTC_TS_CUDA check necessary here? isn't there already an API in LTC which reads this variable and sets some default lazy device field?

**Review by wconstab — APPROVED** [review]
overall lgtm as an incremental improvement! thanks @desertfire

File: lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/aten_ltc_ts_type.cpp
Comment (alanwaketan) [inline lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/aten_ltc_ts_type.cpp:272]: Do you think we can also have a similar reduced story for CPU, like supporting the most common one?

Comment (wconstab) [reply lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/aten_ltc_ts_type.cpp:272]: It'd be good to confirm that the cudnn conv we picked is indeed useful for the nvfuser team since a main motivation for doing the short term fix was to unblock their channels-last work, and also get a read on how soon Joel thinks refactored convolution will be ready.

We probably have bigger fish to fry than supporting more convs manually, but we can discuss the tradeoffs if there is some use case someone wants to enable.

Comment (alanwaketan) [reply lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/aten_ltc_ts_type.cpp:272]: My intent is that if we can have at least one CPU conv that can be used by TorchBench's resnet50, then we can have some early performance numbers.

Comment (desertfire) [reply lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/aten_ltc_ts_type.cpp:272]: CPU has more kernels to be called, conv2d/conv3d, transposed/non-transposed, group/no-group, etc. I don't think it's worth to add it in this temp fix. I think we can focus on GPU for now. Even with our nvfuser friends looking at perf, there are plenty of space and benchmarks left for us to investigate on GPU.

**Comment by byronyi** [comment]
Lowering to TS for convolution_overrideable_backward seems to be broken.

A short repro:

```
import torch
import torch.nn as nn

import lazy_tensor_core

lazy_tensor_core._LAZYC._ltc_init_ts_backend()


data = torch.rand((4, 3, 28, 28))
model = nn.Conv2d(3, 16, kernel_size=(3, 3), padding=1)
output = model(data)
loss = output.sum()
loss.backward()

lazy_loss = model.to('lazy')(data.to('lazy')).sum()
lazy_loss.backward()
```

```
 LTC_TS_CUDA=1 python3 ltc_conv.py
Traceback (most recent call last):
  File "/home/byronyi/tests/ltc_conv.py", line 16, in <module>
    lazy_loss.backward()
  File "/home/byronyi/.virtualenv/ltc/lib/python3.9/site-packages/torch/_tensor.py", line 352, in backward
    torch.autograd.backward(self, gradient, retain_graph, create_graph, inputs=inputs)
  File "/home/byronyi/.virtualenv/ltc/lib/python3.9/site-packages/torch/autograd/__init__.py", line 173, in backward
    Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
RuntimeError: Function ConvolutionOverrideableBackward0 returned an invalid gradient at index 2 - got [4, 16, 28, 28] but expected shape compatible with [16]
```

cc @wconstab @desertfire

**Comment by wconstab** [comment]
@byronyi We have a hacky implementation of convolution at the moment.  We are depending on pytorch core team to refactor convolution operators into `convolution` and `convolution_backward`, at which point we will trace those and won't have to deal with the complexity of many different conv operators.  This work should be completed in the next month if nothing unforeseen comes up.

**Comment by byronyi** [comment]
Changing bias to false bypasses this issue:

```
import torch
import torch.nn as nn

import lazy_tensor_core

lazy_tensor_core._LAZYC._ltc_init_ts_backend()


data = torch.rand((4, 3, 28, 28))
model = nn.Conv2d(3, 16, kernel_size=(3, 3), padding=1, bias=False)
output = model(data)
loss = output.sum()
loss.backward()

lazy_loss = model.to('lazy')(data.to('lazy')).sum()
lazy_loss.backward()
```

From my own lazy prototype one needs to add `grad_bias = grad_output.sum({0, 2, 3});` to CuDNN lowering.

The complete lowering:

```
std::tuple<at::Tensor, at::Tensor, at::Tensor>
convolution_backward_overrideable_cuda(
    const at::Tensor &grad_output, const at::Tensor &input,
    const at::Tensor &weight, at::IntArrayRef stride, at::IntArrayRef padding,
    at::IntArrayRef dilation, bool transposed, at::IntArrayRef output_padding,
    int64_t groups, std::array<bool, 3> output_mask) {
  at::Tensor grad_input, grad_weight, grad_bias;
  if (output_mask[0]) {
    grad_input = at::cudnn_convolution_backward_input(
        input.sizes(), grad_output, weight, padding, stride, dilation, groups,
        /*benchmark=*/true, /*deterministic=*/true, /*allow_tf32=*/true);
  }
  if (output_mask[1]) {
    grad_weight = at::cudnn_convolution_backward_weight(
        weight.sizes(), grad_output, input, padding, stride, dilation, groups,
        /*benchmark=*/true, /*deterministic=*/true, /*allow_tf32=*/true);
  }
  if (output_mask[2]) {
    grad_bias = grad_output.sum({0, 2, 3});
  }
  return {grad_input, grad_weight, grad_bias};
}
```

**Comment by wconstab** [comment]
@byronyi If you want to put a PR up with the fix, I think we can merge it.  (@desertfire any issues with this proposal?)

Otherwise, we'll probably just wait till we get convolution support from pytorch core.

**Pull request merged**
